File: core/src/ru/myitschool/dcrawler/skills/SpearSting.java
package ru.myitschool.dcrawler.skills;

import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.utils.Array;
import ru.myitschool.dcrawler.entities.Entity;
import ru.myitschool.dcrawler.math.DiceAction;
import ru.myitschool.dcrawler.math.MathAction;
import ru.myitschool.dcrawler.skills.action.Action;
import ru.myitschool.dcrawler.skills.action.SimpleAttackAction;
import ru.myitschool.dcrawler.skills.patterns.FloorWaveTargetPattern;
import ru.myitschool.dcrawler.skills.patterns.TargetPattern;
import ru.myitschool.dcrawler.skills.targeting.TargetDisplayer;
import ru.myitschool.dcrawler.skills.targeting.WaveDisplayer;

/**
 * Created by Voyager on 02.12.2017.
 */
public class SpearSting extends Skill {

    MathAction attackAction = new DiceAction(10);

    public SpearSting(Entity doer) {
        super(doer);
        addPlayContainer().getEntityPlay().addAction(new SimpleAttackAction(this, attackAction));
        addPlayContainer();
    }

    @Override
    public void maintainDisplayers(Array<TargetDisplayer> displayers) {
        displayers.add(new WaveDisplayer());
    }

    @Override
    public TargetPattern getPattern() {
        return new FloorWaveTargetPattern(this);
    }

    @Override
    public Texture getIcon() {
        return new Texture("spear_sting.png");
    }

    @Override
    public String getName() {
        return "Spear sting";
    }

    @Override
    public String getDescription() {
        return "You attacks all enemies that stands in two cells in front of you and deal " + attackAction.getDescription() + " damage";
    }

    @Override
    public int getSkillAccuracyBonus() {
        return 0;
    }

    @Override
    public int getRange() {
        return 2;
    }

    @Override
    public int getDistanceMin() {
        return 1;
    }

    @Override
    public int getDistanceMax() {
        return 1;
    }

    @Override
    public int getTargetCountMax() {
        return 1;
    }

    @Override
    public int getCooldownMax() {
        return 0;
    }

    @Override
    public int getType() {
        return SKILL_TYPE_AT_WILL;
    }

    @Override
    public int getTargetType() {
        return SKILL_TARGET_TYPE_FLOOR_WAVE;
    }

    @Override
    public boolean isCheckAllTargets() {
        return false;
    }

    @Override
    public boolean isMarkEverything() {
        return false;
    }

    @Override
    public boolean isMark() {
        return true;
    }

    @Override
    public boolean isObstruct() {
        return true;
    }

    @Override
    public boolean isWallTargets() {
        return false;
    }
}

File: core/src/com/wolg_vlad/dcrawler/entities/skills/SpearSting.kt
package com.wolg_vlad.dcrawler.entities.skills

import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.utils.Array
import com.wolg_vlad.dcrawler.entities.Entity
import com.wolg_vlad.dcrawler.entities.skills.action.SimpleAttackAction
import com.wolg_vlad.dcrawler.entities.skills.patterns.FloorWaveTargetPattern
import com.wolg_vlad.dcrawler.entities.skills.patterns.TargetPattern
import com.wolg_vlad.dcrawler.entities.skills.targeting.TargetDisplayer
import com.wolg_vlad.dcrawler.entities.skills.targeting.WaveDisplayer
import com.wolg_vlad.dcrawler.math.DiceAction
import com.wolg_vlad.dcrawler.math.MathAction

/**
 * Created by Voyager on 02.12.2017.
 */
class SpearSting(doer: Entity) : Skill(doer) {
    var attackAction: MathAction = DiceAction(10)

    override val displayers: kotlin.Array<TargetDisplayer> = arrayOf(WaveDisplayer())
    override val pattern: TargetPattern = FloorWaveTargetPattern(this)
    override val icon: Texture = Texture("spear_sting.png")
    override val name: String = "Spear sting"
    override val description: String = "You attacks all enemies that stands in two cells in front of you " +
            "and deal ${attackAction.description} damage"
    override val skillAccuracyBonus: Int = 0
    override val range: Int = 2
    override val distanceMin: Int = 1
    override val distanceMax: Int = 1
    override val targetCountMax: Int = 1
    override val cooldownMax: Int = 0
    override val type: Int = SKILL_TYPE_AT_WILL
    override val targetType: Int = SKILL_TARGET_TYPE_FLOOR_WAVE
    override val isCheckAllTargets: Boolean = false
    override val isMarkEverything: Boolean = false
    override val isMark: Boolean = true
    override val isObstruct: Boolean = true
    override val isWallTargets: Boolean = false

    init {
        addPlayContainer().entityPlay.addAction(SimpleAttackAction(this, attackAction))
    }
}